nisystemlink/clients/dataframe/_data_frame_client.py

```diff
@@ -250,18 +256,181 @@ def get_table_data(
+                    parameter.
+                * Iterable[pyarrow.RecordBatch]: Streamed as Arrow IPC. ``end_of_data`` (if provided)
+                    is sent as a query parameter. If the iterator yields no batches, it is treated
+                    as ``None`` and requires ``end_of_data``.
+                * None: ``end_of_data`` must be provided; sends JSON containing only the
+                    ``endOfData`` flag.
+            end_of_data: Whether additional rows may be appended in future requests. Required when
+                    ``data`` is ``None`` or the RecordBatch iterator is empty; must be omitted when
+                    passing an ``AppendTableDataRequest`` (include it inside that model instead).
 
         Raises:
-            ApiException: if unable to communicate with the DataFrame Service
-                or provided an invalid argument.
+            ValueError: If parameter constraints are violated.
+            ApiException: If unable to communicate with the DataFrame Service or an
+                invalid argument is provided.
         """
-        ...
+        if isinstance(data, models.AppendTableDataRequest):
+            if end_of_data is not None:
+                raise ValueError(
+                    "end_of_data must not be provided separately when passing an AppendTableDataRequest."
+                )
+            self._append_table_data_json(id, data)
+            return
+
+        if isinstance(data, models.DataFrame):
+            if end_of_data is None:
+                request_model = models.AppendTableDataRequest(frame=data)
+            else:
+                request_model = models.AppendTableDataRequest(
+                    frame=data, end_of_data=end_of_data
+                )
+            self._append_table_data_json(id, request_model)
+            return
+
+        if pa is not None and isinstance(data, pa.RecordBatch):
+            data = [data]
+
+        if isinstance(data, Iterable):
+            iterator = iter(data)
+            try:
+                first_batch = next(iterator)
+            except StopIteration:
+                if end_of_data is None:
+                    raise ValueError(
+                        "end_of_data must be provided when data iterator is empty."
+                    )
+                self._append_table_data_json(
+                    id,
+                    models.AppendTableDataRequest(end_of_data=end_of_data),
+                )
+                return
+
+            if pa is None:
+                raise RuntimeError(
+                    "pyarrow is not installed. Install to stream RecordBatches."
+                )
+
+            if not isinstance(first_batch, pa.RecordBatch):
+                raise ValueError(
+                    "Iterable provided to data must yield pyarrow.RecordBatch objects."
+                )
+
+            def _generate_body() -> Iterable[memoryview]:
+                data_iter = iter(data)
```

Comment:
This will iterate a second time, which may lead to unexpected behavior if clients are generating their record batches. We should be able to pass in the `iterator` and `first_batch` from above and use those.